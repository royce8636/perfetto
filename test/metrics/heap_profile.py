#!/usr/bin/python
# Copyright (C) 2018 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from os import sys, path
sys.path.append(path.dirname(path.dirname(path.abspath(__file__))))
import synth_common

class ProfileBuilder(object):
  def __init__(self, pid, profile_packet):
    self.profile_packet = profile_packet
    self.profile_packet.mappings.add().iid = 1 # Dummy mapping
    self.process_dump = profile_packet.process_dumps.add()
    self.process_dump.pid = pid
    self.interned = {}

  def add_sample(self, callstack_id, allocs, frees):
    sample = self.process_dump.samples.add()
    sample.callstack_id = callstack_id
    sample.self_allocated = 1000 * allocs
    sample.self_freed = 1000 * frees
    sample.alloc_count = allocs
    sample.free_count = frees

  def add_callstack(self, id, frames):
    callstack = self.profile_packet.callstacks.add()
    callstack.iid = id
    fid = 1
    for frame in frames:
      callstack.frame_ids.append(fid)
      frame_proto = self.profile_packet.frames.add()
      frame_proto.iid = fid
      frame_proto.function_name_id = self._intern(frame)
      frame_proto.mapping_id = 1
      fid += 1

  def _intern(self, value):
    if value in self.interned:
      return self.interned[value]
    next_id = len(self.interned) + 1
    self.interned[value] = next_id
    s = self.profile_packet.strings.add()
    s.iid = next_id
    s.str = value
    return next_id

trace = synth_common.create_trace()
trace.add_process_tree_packet()
trace.add_process(1, 0, 'init')
trace.add_process(pid=2, ppid=1, cmdline='system_server')

pbuilder = ProfileBuilder(pid=2, profile_packet=trace.add_profile_packet(ts=10))
pbuilder.add_sample(callstack_id=1, allocs=2, frees=1)
pbuilder.add_callstack(id=1, frames=['f1', 'f2', 'f3'])

print(trace.trace.SerializeToString())
