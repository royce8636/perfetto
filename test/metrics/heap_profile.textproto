packet {
  process_tree {
    processes {
      pid: 1
      ppid: 0
      cmdline: "init"
    }
    processes {
      pid: 2
      ppid: 1
      cmdline: "system_server"
    }
  }
}
packet {
  trusted_packet_sequence_id: 999
  timestamp: 10
  profile_packet {
    strings {
      iid: 1
      str: "f1"
    }
    strings {
      iid: 2
      str: "f2"
    }
    strings {
      iid: 3
      str: "f3"
    }
    strings {
      iid: 4
      str: "liblib.so"
    }
    strings {
      iid: 5
      str: "build-id"
    }
    frames {
      iid: 1
      function_name_id: 1
      mapping_id: 1
      rel_pc: 0x1000
    }
    frames {
      iid: 2
      function_name_id: 2
      mapping_id: 1
      rel_pc: 0x2000
    }
    frames {
      iid: 3
      function_name_id: 3
      mapping_id: 1
      rel_pc: 0x3000
    }
    callstacks {
      iid: 1
      frame_ids: 1
      frame_ids: 2
      frame_ids: 3
    }
    mappings {
      iid: 1
      path_string_ids: 4
      build_id: 5
    }
    process_dumps {
      pid: 2
      samples {
        callstack_id: 1
        self_allocated: 2000
        self_freed: 1000
        alloc_count: 2
        free_count: 1
      }
    }
  }
}
# Add some symbolization packets
packet {
  module_symbols {
    path: "/liblib.so"
    build_id: "build-id"
    address_symbols {
      address: 0x3000
      lines {
        function_name: "symbolized f3"
        source_file_name: "f3.cc"
        line_number: 33
      }
    }
  }
}
