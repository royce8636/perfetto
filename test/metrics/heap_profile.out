heap_profile {
  profile {
    process_name: "system_server"
    callsite_stats {
      callsite {
        frame {
          name: "f3"
        }
        frame {
          name: "f2"
        }
        frame {
          name: "f1"
        }
      }
      allocs_self_count: 1
      allocs_self_bytes: 1000
    }
  }
}
